lib/Expression_gen.ml

```diff
@@ -1,262 +1,313 @@
+            else pp_ordinary_fn ppf f es ) ]
+  in
+  let read_data_or_param ppf es =
+    let i_or_r =
+      match ut with
+      | Ast.UInt -> "i"
+      | UReal -> "r"
+      | UVector | URowVector | UMatrix | UArray _
+       |UFun (_, _)
+       |UMathLibraryFunction ->
+          raise_s [%message "Can't ReadData of " (ut : unsizedtype)]
+    in
+    pf ppf "context__.vals_%s(%a)" i_or_r pp_expr (List.hd_exn es)
+  in
+  let mir_special_functions =
+    Map.Poly.of_alist_exn
+      [ (fn_length, fun ppf es -> pp_unary ppf "length(%a)" es)
+      ; ( fn_make_array
+        , fun ppf es -> pf ppf "{%a}" (list ~sep:comma pp_expr) es )
+      ; (fn_read_data, read_data_or_param)
+      ; (fn_read_param, read_data_or_param)
+      ; (fn_constrain, pp_constrain_funapp "constrain")
+      ; (fn_unconstrain, pp_constrain_funapp "unconstrain")
+      (* XXX Fill in the rest of the functions at the bottom of Mir.ml *)
+       ]
+  in
+  let default ppf es = pp_ordinary_fn ppf (stan_namespace_qualify f) es in
+  let pp_fn =
+    [operator_functions; misc_special_math_fns; mir_special_functions]
+    |> List.find_map ~f:(fun m -> Map.Poly.find m f)
+    |> Option.value ~default
+  in
+  pp_fn ppf es
+
+(* XXX actually, for params we have to combine read and constrain into one funapp *)
+and pp_constrain_funapp constrain_or_un_str ppf = function
+  | var
+    :: {texpr= Lit (Str, constraint_flavor); _}
+       :: {texpr= Lit (Str, base_type); _} :: dims ->
+      pf ppf "%s_%s_%s(@[<hov>%a@])" base_type constraint_flavor
+        constrain_or_un_str (list ~sep:comma pp_expr) (var :: dims)
+  | es -> raise_s [%message "Bad constraint " (es : expr_typed_located list)]
+
+and pp_ordinary_fn ppf f es =
+  let extra_args = gen_extra_fun_args f in
+  let sep = if List.is_empty extra_args then "" else ", " in
+  pf ppf "%s(@[<hov>%a%s@])" f (list ~sep:comma pp_expr) es
+    (sep ^ String.concat ~sep:", " extra_args)
+
+and pp_expr ppf (e : expr_typed_located) =
   match e.texpr with
-  | Var s -> s
-  | Lit (Str, s) -> sprintf "%S" s
-  | Lit (_, s) -> s
-  | FunApp (f, es) -> gen_fun_app f es
+  | Var s -> pf ppf "%s" s
+  | Lit (Str, s) -> pf ppf "%S" s
+  | Lit (_, s) -> pf ppf "%s" s
+  | FunApp (f, es) -> gen_fun_app ppf e.texpr_type f es
+  | And (e1, e2) -> pp_logical_op ppf "&&" e1 e2
+  | Or (e1, e2) -> pp_logical_op ppf "||" e1 e2
   | TernaryIf (ec, et, ef) ->
-    if types_match et ef then
-      sprintf "(%s ? %s : %s)" (gen_expr ec) (gen_expr et) (gen_expr ef)
-    else
-      sprintf
-        "(%s ? stan::math::promote_scalar<%s>(%s) : \
-         stan::math::promote_scalar<%s>(%s)"
-        (gen_expr ec)
-        (gen_type e)
-        (gen_expr et)
-        (gen_type e)
-        (gen_expr ef)
+      if types_match et ef then
+        pf ppf "(%a ? %a : %a)" pp_expr ec pp_expr et pp_expr ef
+      else
+        pf ppf
+          "(%a ? stan::math::promote_scalar<%a>(%a) : \
```

Comment:
This is the kind of thing we can break if we want to generate

```
(a
 ? b
 : c)
```

So doesn't that require boxing control.  

Same with `(a || b)` ---the `||` can go on the next line.